mono/sgen/sgen-minor-scan-object.h

```diff
@@ -104,16 +111,31 @@ SERIAL_SCAN_PTR_FIELD (GCObject *full_object, GCObject **ptr, SgenGrayQueue *que
 static gboolean
 SERIAL_DRAIN_GRAY_STACK (SgenGrayQueue *queue)
 {
-        for (;;) {
-                GCObject *obj;
-                SgenDescriptor desc;
+#ifdef SGEN_SIMPLE_PAR_NURSERY
+	int i;
+	/*
+	 * We do bounded iteration so we can switch to optimized context
+	 * when we are the last worker remaining.
+	 */
+	for (i = 0; i < 32; i++) {
```

Comment:
This is what we were using for the concurrent major since the beginnings (in order to support stopping it). The purpose of this is to be fairly responsive to stopping. This area will also need attention in the future since we scan los objects as a whole, so these loops don't guarantee much.